tests: fix race conditions in integration tests

pbrezina opened the pull request:
The amount of failure in our PR CI increased because the number
of tested distribution grew. Lets add more time for SSSD to
pick up changes to avoid these failures.

alexey-tikhonov commented:
Probably, those changes need better justification taking into account [3463](https://pagure.io/SSSD/sssd/issue/3463) and #345...

pbrezina commented:
I will add some, once I know it fixes the issue (I'm pretty confident about it). Lets wait for the tests to finish.

pbrezina commented:
First run succeeded. I added some more explanation to the commits and let's see if we get all green result the seconds time.

pbrezina commented:
Second time succeeded. I altered the commit message little bit and add a comment to enumeration tests.

pbrezina commented:
retest this please

pbrezina commented:
retest this please

pbrezina commented:
All tests passed third time. The failure in CentOS CI is not related: it is some infrastructure error that started to happen today, see [logs](https://ci.centos.org/job/sssd-CentOS7/2411/console).

pbrezina commented:
@alexey-tikhonov Can you review this please? I would like to merge it soon so we have green tests.

alexey-tikhonov commented:
In regards of `test_files_provider.py`

There is common pattern like:
```
def check_group(exp_group, delay=3.0):
    if delay > 0:
        time.sleep(delay)

    res, found_group = sssd_getgrnam_sync(exp_group["name"])
    assert res == NssReturnCode.SUCCESS
    assert found_group == exp_group
```
but  `sssd_get..._sync()` calls `poll_canary()` which has its own "sleep" (20*0.1 sec currently).

This seems to be original design (not introduced by this patch).

`poll_canary()` only awaits if result is `NOTFOUND` and has nice explanation of its legitimacy in the comment.
But other sleeps are unconditional and its intention is not that clear. Some explanation is given in the original commit message:
```
    during testing especially on slow machines, it became apparent that
    sometimes the inotify message arrives later than the test would check
    for the changed entries. Therefore, the check would query the NSS
    responder even before the sss-files domain was invalidated.
```
So it seems intention is to cover case when files content was *modified* (not added/deleted). Like here:
```
    modgroup = dict(GROUP1)
    modgroup['name'] = 'group1_mod'
    add_group_with_canary.groupmod(old_name=GROUP1["name"], **modgroup)

    check_group(modgroup)
```
But IMO this hides a bug in a files provider: sssd *does* return stale data in this case. And I am not sure glossing over this bug in the test is a good idea. 

Some relevant to this problem thoughts can be found here: https://bugzilla.redhat.com/show_bug.cgi?id=1693379#c6 (what we hit here IMO is more or less the same race as mentioned there)

Given above, I would probably opt to:
 - increase threshold in `poll_canary()`;
 - get rid of all other sleeps
 - and disable/mark "expected to fail" tests that currently expose bug in files_provider & open ticket for files_provder.

OTOH I still see some value in those tests even with sleeps. It just looks bad...

alexey-tikhonov commented:
In regards of `test_enumeration.py`

 - I am not sure Nyquist frequency is relevant. I agree that test should wait longer than enumeration period to allow reliable update, but I am not convinced this should be twice longer. Could you please elaborate?
 - Do You understand the reason behind `time.sleep(INTERACTIVE_TIMEOUT/2)` in the beginning of each test before any action is taken? This seems just senseless for me. This is especially weird because this is timeout that was changed by #345 and this change was reported to help, but I hardly understand how this could have reliable affect.

So basically I agree with approach - to make INTERACTIVE_TIMEOUT greater than ENUMERATION_TIMEOUT - but would offer other values and to remove `time.sleep(INTERACTIVE_TIMEOUT/2)`

pbrezina commented:
Enumeration is run periodically. It is scheduled during SSSD startup and then it runs every four seconds. We do not know precisely when the enumeration refresh is triggered so sleeping for four seconds is not enough, we have to sleep longer time to adapt for delays (it is not scheduled on precise time because of other sssd operation, it takes some time to finish so each run moves it further away from exact 4 seconds period, cpu scheduler, context switches, ...). I agree that just little bit longer timeout may work as well. However we can be sure when using twice the period because we know that enumeration was indeed run at least once.

I think the `/2`  sleep is to wait until the first enumeration is finished. But we would have to ask the original author.

About files provider tests: I checked if there is a bug that you mentioned and no. There does not seem to be any bug, though the behavior is not the same as the comment says in the tests. If the files provider is updating cache, then getpw/gr calls will block (see flow around `sss_dp_account_files_params`). So there should be no need for polling nor any kind of sleep. So there is something else going on. I'm disabling memory cache in tests, lets see if this helps.

alexey-tikhonov commented:
> Enumeration is run periodically. It is scheduled during SSSD startup and then it runs every four seconds. We do not know precisely when the enumeration refresh is triggered so sleeping for four seconds is not enough, we have to sleep longer time to adapt for delays (it is not scheduled on precise time because of other sssd operation, it takes some time to finish so each run moves it further away from exact 4 seconds period, cpu scheduler, context switches, ...). 

Right, I agree.

> I agree that just little bit longer timeout may work as well. However we can be sure when using twice the period because we know that enumeration was indeed run at least once.

We have to have 
```
INTERACTIVE_TIMEOUT > "REAL"_ENUMERATION_TIMEOUT,
```
where REAL_ENUMERATION_TIMEOUT = ENUMERATION_TIMEOUT + some non deterministic `delay`.
But:
1) strictly speaking there is no guarantee `delay` < ENUMERATION_TIMEOUT (but otherwise of course would be insane environment)
 2) I think ENUMERATION_TIMEOUT as an estimation of this `delay` is too much. This would not be a problem if intg-tests would not consume that long time to run...

tl,dr: I'm not categorically against estimation of  REAL_ENUMERATION_TIMEOUT as 2*ENUMERATION_TIMEOUT.
But IMO this has nothing to do with Nyquist frequency and this is still arbitrary  (just "good enough with a very good margin") value.

> I think the `/2` sleep is to wait until the first enumeration is finished. 

To be sure that there is really no entry in ldap before test adds it? Makes sense.
But:
1) why then it is /2? As discussed above that's not enough even with this patch. And it was clearly too small for this purpose before, so intention of this timeout is still not clear to me.
2) do we really need this in the beginning of every test? I am not sure if sssd restarted in every test or once for the whole module?

It is 5*(INTERACTIVE_TIMEOUT/2) = 20 seconds for each run (or twice if we remove /2). Doesn't sound like too much but things add up. And CI runs are really slow...

alexey-tikhonov commented:
> About files provider tests: I checked if there is a bug that you mentioned and no. There does not seem to be any bug, though the behavior is not the same as the comment says in the tests. If the files provider is updating cache, then getpw/gr calls will block (see flow around `sss_dp_account_files_params`).

Hm... I thought files_provider during update tells nss_responder to "stop serve" files domain so that responder returns NOT_FOUND.

> So there should be no need for polling nor any kind of sleep.

Sorry if I was not clear enough.
I was talking about different bug/race: when updated data are requested *before* files_provider even started update.
1) edit actual files (/etc/passwd...)
2) files_provider notices the change (via inotify) and starts internal update
3) updated is finished, provider is ready

You are talking about request during (2).
I am talking about request between (1) and (2). I believe this is a reason of additional sleeps and hides a bug.

pbrezina commented:
> tl,dr: I'm not categorically against estimation of REAL_ENUMERATION_TIMEOUT as 2*ENUMERATION_TIMEOUT.
> But IMO this has nothing to do with Nyquist frequency and this is still arbitrary (just "good enough with a very good margin") value.

You did not give me any argument to convince me otherwise. There is no point spending time on this as long as the tests works so let's agree to disagree. I can remove this name from comments and code if you feel strongly about it.

> > I think the `/2` sleep is to wait until the first enumeration is finished.
> 
> To be sure that there is really no entry in ldap before test adds it? Makes sense.
> But:
> 
>     1. why then it is /2? As discussed above that's not enough even with this patch. And it was clearly too small for this purpose before, so intention of this timeout is still not clear to me.
> 
>     2. do we really need this in the beginning of every test? I am not sure if sssd restarted in every test or once for the whole module?

I really do not know. @spbnick Git blame says you are the original author of these tests. Do you remember why you used `time.sleep(INTERACTIVE_TIMEOUT/2)` at the beginning of each enumeration test in `test_enumeration.py`?

> It is 5*(INTERACTIVE_TIMEOUT/2) = 20 seconds for each run (or twice if we remove /2). Doesn't sound like too much but things add up. And CI runs are really slow...

I don't care that it will be five minutes slower if it runs correctly.

alexey-tikhonov commented:
> Sorry if I was not clear enough.
> I was talking about different bug/race: when updated data are requested _before_ files_provider even started update.
>     1. edit actual files (/etc/passwd...)
>     2. files_provider notices the change (via inotify) and starts internal update
>     3. updated is finished, provider is ready
> 
> You are talking about request during (2).
> I am talking about request between (1) and (2). I believe this is a reason of additional sleeps and hides a bug.

And yes, mem-cache actually adds complexity to this issue.

alexey-tikhonov commented:
> > tl,dr: I'm not categorically against estimation of REAL_ENUMERATION_TIMEOUT as 2*ENUMERATION_TIMEOUT.
> > But IMO this has nothing to do with Nyquist frequency and this is still arbitrary (just "good enough with a very good margin") value.
> 
> You did not give me any argument to convince me otherwise.

Hm... Not sure how could I convince you that I just do not understand how it is relevant :) (which doesn't necessarily mean it is not, I really might miss something)

> I can remove this name from comments and code if you feel strongly about it.

I would opt to either (preferably) explain why "*2" is sort of special/"guaranteed" safe value (i.e. how sampling rate/high freq applicable here) or stick with "good enough margin" explanation. I am fine with both.

> I don't care that it will be five minutes slower if it runs correctly.

Then we could just change /2 -> 1 and push test_enumeration patch. 
Case of files provider is more complicated IMO.

spbnick commented:
@pbrezina, Hi Pavel :)
I don't really remember, sorry, but looking at the code now it might be to avoid racing with SSSD initialization, while making the change before the timeout expires. I.e. to have the database change between the SSSD initialization and the first timeout, perhaps?

alexey-tikhonov commented:
> @pbrezina, Hi Pavel :)
> I don't really remember, sorry, but looking at the code now it might be to avoid racing with SSSD initialization, while making the change before the timeout expires. I.e. to have the database change between the SSSD initialization and the first timeout, perhaps?

If value of this timeout has nothing to do with enumeration period I would prefer to replace/rename `INTERACTIVE_TIMEOUT/2` as this is a little bit misleading.

pbrezina commented:
Thank you both for your time. I amended in-code comments and commit messages.

1) About files provider.

You are right that the sleep is suppose to eliminated race condition between touching the files and triggering inotify callback so it has to stay. There is nothing we can do about it. There are ways to workaround it of course (e.g. signalling domain state) but those does not have real use case so there is no point doing that.

`poll_canary` has no meaning as we currently block when refreshing the cache. We can either remove it or report a bug against sssd that it behaves incorrectly. @jhrozek When files provider is updating cache in inotify callback (i.e. domain is inconsistent) should nss responder A) return not found and let client look in files database, or B) block until the cache is updated? Currently B is true but tests says A should be done.

In fact if SSSD uses A, I have some suggestions for [aforementioned bugzilla](https://bugzilla.redhat.com/show_bug.cgi?id=1693379).

2) About enumeration tests

I kept mentioning Nyquist in commit message because in my opinion it is related and it is why I used twice the value, but I removed it from the code. So I hope we are both satisfied this way.

I also removed the `/2` sleep as nobody sees its value and tests actually pass without it successfully as well (locally for now, lets wait for PR CI result).

I would like to get these patches merged so the tests are all green again. And do any further improvements in different PR if needed/requested.

alexey-tikhonov commented:
F29 failed `test_enumeration.py::test_add_remove_membership_rfc2307`...

alexey-tikhonov commented:
> pbrezina force-pushed the pbrezina:fixtests

Looks like change of the value of this `/2` timeout is what actually fixes race. And this is exactly the approach that was used in #345...

pbrezina commented:
Ok, lets try it with Fabiano's patch.

pbrezina commented:
Works. Let try another run.

pbrezina commented:
All green. One more time.

pbrezina commented:
There was an infrastructure failure, not related to tests. Running again.

alexey-tikhonov commented:
Probably it makes sense to split this PR into two different?

alexey-tikhonov commented:
Btw, I still think it makes sense to have `INTERACTIVE_TIMEOUT > ENUMERATION_TIMEOUT + delay`. Even if currently CI runs fine with (equal timeouts & `/2`->`/1`)

pbrezina commented:
Tests works now. If we will see more failures we can always push another patch. There is no reason to spend more time on this. So I kindly as you to finish this unless you strongly disagree with current patches.

alexey-tikhonov commented:
First of all:
 - I understand importance of "green" CI;
 - those patches make CI "green".

So thank you for putting an effort in attempt to resolve this issue (that lasts quite long time).

But:

I. `test_files_provider.py`

1)
> You are right that the sleep is suppose to eliminated race condition between touching the files and triggering inotify callback so it has to stay.

I see this as a bug in the SSSD (libnss_files.so doesn't have such an issue).
And I think hiding this bug in a test is not the best way to deal with it.

Probably I am wrong and this behavior of SSSD:files_provider can be justified.
But I am not sure if PR comments is proper place to do such justification.

From my point of view it would be preferable to:
1) mark parts of the test that fail due to described race as "expected to fail" (to make CI usable)
2) open issue against files_provider, triage it properly
3) depending on the results of (2) either wait for fix or allow extended timeout in the test (this patch)

Maybe this is way too pedantic. Maybe comments that you added to the test code, explaining that test is only intended to check state *after* inotify callback run, are "good enough".
I would really like to listen opinion of other maintainer...

> `poll_canary` has no meaning as we currently block when refreshing the cache. We can either remove it or report a bug against sssd that it behaves incorrectly.

I agree. This is another known issue with this test. I would await @jhrozek response first.
But probably you are right and this can be a subject of different PR.

II. `test_enumeration.py`

1) Given your [nice explanation](https://github.com/SSSD/sssd/pull/947#issuecomment-559440211)  I am convinced `INTERACTIVE_TIMEOUT` should be greater than `ENUMERATION_TIMEOUT + delay`. Recent patchset misses this.
2) I am afraid I can't ack change of timeout meaning of which I do not understand (I mean `INTERACTIVE_TIMEOUT/2`). But of course I can investigate it.


Based on the above, I would like other maintainers to provide their opinion.

pbrezina commented:
This is exactly the reason why it has not been fixed for two+ years. You remind me Lukas so much (in both good and bad way) :-) @sumit-bose @thalman @mzidek-rh can any of you chime in?

pbrezina commented:
BTW I am not opposing your arguments, it is just not worth the time (and then again I could have already written a patch instead of commenting). So feel free to take over and submit different patch. I just want to have successful tests so PR CI is meaningful.

pbrezina commented:
> I see this as a bug in the SSSD (libnss_files.so doesn't have such an issue).
> And I think hiding this bug in a test is not the best way to deal with it.

Its not a bug, it is a feature because SSSD is notified about the changes through inotify instead of reading the file. It should be remmedied - yes. File a ticket for it. When changed, new tests for this change will be written. But currently, it is a feature that tests need to account for.

> Probably I am wrong and this behavior of SSSD:files_provider can be justified.
> But I am not sure if PR comments is proper place to do such justification.
> 
> From my point of view it would be preferable to:
> 
>     1. mark parts of the test that fail due to described race as "expected to fail" (to make CI usable)

That means tests are unstable, unusable and we may as well remove them because they do not test anything.

>     2. open issue against files_provider, triage it properly

See above.

>     3. depending on the results of (2) either wait for fix or allow extended timeout in the test (this patch)

We are testing current code, not future changes.

alexey-tikhonov commented:
I. `test_files_provider.py`

> `poll_canary` has no meaning as we currently block when refreshing the cache. We can either remove it or report a bug against sssd that it behaves incorrectly. @jhrozek When files provider is updating cache in inotify callback (i.e. domain is inconsistent) should nss responder A) return not found and let client look in files database, or B) block until the cache is updated? Currently B is true but tests says A should be done.

Actually... Are you sure "B is true"?
I just tested with a patch of `providers/files/files_ops.c:sf_passwd_cb()`:
`+    while (1) { usleep(100000); }`
right before `sf_enum_files()` and `id` works fine (according to your statement this should block).

I checked with `strace`:
```
openat(AT_FDCWD, "/lib64/libnss_sss.so.2", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/var/lib/sss/mc/passwd", O_RDONLY|O_CLOEXEC) = 3
connect(4, {sa_family=AF_UNIX, sun_path="/var/lib/sss/pipes/nss"}, 110) = 0
openat(AT_FDCWD, "/lib64/libnss_files.so.2", O_RDONLY|O_CLOEXEC) = 5
```
so it falls back to `files`.

Probably reason is call to `dp_sbus_domain_inconsistent()` in the beginning of `sf_passwd_cb()`?

So I think this part:
> > `poll_canary` has no meaning as we currently block when refreshing the cache. We can either remove it or report a bug against sssd that it behaves incorrectly.
> I agree. This is another known issue with this test.

is void.


> This is exactly the reason why it has not been fixed for two+ years.

I am sorry if I am not constructive.

> > I see this as a bug in the SSSD (libnss_files.so doesn't have such an issue).
> > And I think hiding this bug in a test is not the best way to deal with it.
> 
> Its not a bug, it is a feature

One can call it "design flaw". But regression of user experience at comparison with default libnss_files is not a "feature" in any way.

> It should be remmedied - yes. File a ticket for it.

[4125](https://pagure.io/SSSD/sssd/issue/4125)


> When changed, new tests for this change will be written. 
> ...
> We are testing current code, not future changes.

Ok, this point of view makes sense.

For what it's worth: here is my ACK for `test_files_provider.py` patch.


II. `test_enumeration.py`

I didn't change my stance here.

> So feel free to take over and submit different patch.

Lets see what other will say. If your patch will not be merged I will try my best to investigate issue and explain `INTERACTIVE_TIMEOUT/2`

sumit-bose commented:
Hi,

about the race between adding e.g. the user to /etc/passwd and getting inotify processed.

Isn't this similar to why calling `sss_cache` was added `useradd` and other utilities from the shadow-utils package? Have you tried if instead of increasing the delay calling `sss_cache -E` would work as well?

If I see it correctly the test just adds a new line to /etc/passwd and then tries to lookup the freshly added user. I think whenever a cache is added in between this use case can only be covered if the cache is explicitly invalidated after adding the new line (shadow-utils is also calling `nscd -i ...` is `nscd` is used). Using inotify helps if invalidating the cache was forgotten but it is only best effort.

bye,
Sumit

alexey-tikhonov commented:
> about the race between adding e.g. the user to /etc/passwd and getting inotify processed.
> 
> Isn't this similar to why calling `sss_cache` was added `useradd` and other utilities from the shadow-utils package? Have you tried if instead of increasing the delay calling `sss_cache -E` would work as well?

If i understand correctly `sss_cache` will reset mem-cache and re-write timestamps in sysdb cache but that's not enough because files_provider ignores timestamps in sysdb cache. So if inotify event have not been processed yet then files_provider still will return stale data from sysdb cache. I wrote [about it](https://bugzilla.redhat.com/show_bug.cgi?id=1693379#c6) (item 2) )

> I think whenever a cache is added in between this use case can only be covered if the cache is explicitly invalidated after adding the new line

This is true in the limits of current design of files_provider.

But the problem is there are no "ready" means to invalidate sysdb cache for files_provider as it ignores timestamp (again, if I understand correctly)

To this extent I agree with @pbrezina that if case is expected to test current design then `sleep` is ok.

pbrezina commented:
> I. `test_files_provider.py`
> 
> > `poll_canary` has no meaning as we currently block when refreshing the cache. We can either remove it or report a bug against sssd that it behaves incorrectly. @jhrozek When files provider is updating cache in inotify callback (i.e. domain is inconsistent) should nss responder A) return not found and let client look in files database, or B) block until the cache is updated? Currently B is true but tests says A should be done.
> 
> Actually... Are you sure "B is true"?
> I just tested with a patch of `providers/files/files_ops.c:sf_passwd_cb()`:
> `+ while (1) { usleep(100000); }`
> right before `sf_enum_files()` and `id` works fine (according to your statement this should block).

I'm sure. See:
* https://github.com/SSSD/sssd/blob/master/src/responder/common/responder_dp.c#L44
* https://github.com/SSSD/sssd/blob/master/src/providers/files/files_id.c#L132
* https://github.com/SSSD/sssd/blob/master/src/providers/files/files_ops.c#L841

> I checked with `strace`:
> 
> ```
> openat(AT_FDCWD, "/lib64/libnss_sss.so.2", O_RDONLY|O_CLOEXEC) = 3
> openat(AT_FDCWD, "/var/lib/sss/mc/passwd", O_RDONLY|O_CLOEXEC) = 3
> connect(4, {sa_family=AF_UNIX, sun_path="/var/lib/sss/pipes/nss"}, 110) = 0
> openat(AT_FDCWD, "/lib64/libnss_files.so.2", O_RDONLY|O_CLOEXEC) = 5
> ```

Because you used `usleep` at wrong place. See, D-Bus communication is asynchronous so once you put a sleep somewhere, the process can not receive nor send any message. In this case, even though `dp_sbus_domain_inconsistent` was called, the message was not send because you immediately go to sleep. Therefore, the nss responder will answer with what it has in cache. You will see this in logs.

I'm not sure why it continues with nss_files, I am not this much familiar with nss internals.

To reproduce it correctly, use rather this diff:

```diff
--- a/src/providers/files/files_ops.c
+++ b/src/providers/files/files_ops.c
@@ -836,9 +836,9 @@ static int sf_passwd_cb(const char *filename, uint32_t flags, void *pvt)
 
     ret = EOK;
 done:
-    id_ctx->updating_passwd = false;
+    // id_ctx->updating_passwd = false;
     sf_cb_done(id_ctx);
-    files_account_info_finished(id_ctx, BE_REQ_USER, ret);
+    // files_account_info_finished(id_ctx, BE_REQ_USER, ret);
     return ret;
 }
```

```
[root /dev/shm/sssd]# id pbrezina@files
uid=1000(pbrezina) gid=1000(pbrezina) groups=1000(pbrezina),10(wheel)
[root /dev/shm/sssd]# touch /etc/passwd
[root /dev/shm/sssd]# sss_cache -E
[root /dev/shm/sssd]# id pbrezina@files
... after a long timeout
id: ‘pbrezina@files’: no such user


[root /dev/shm/sssd]# id pbrezina
uid=1000(pbrezina) gid=1000(pbrezina) groups=1000(pbrezina),10(wheel)
[root /dev/shm/sssd]# touch /etc/passwd
[root /dev/shm/sssd]# sss_cache -E
[root /dev/shm/sssd]# id pbrezina
... after a long timeout
uid=1000(pbrezina) gid=1000(pbrezina) groups=1000(pbrezina),10(wheel)
```

pbrezina commented:
You actually don't even need to call `sss_cache -E` because files provider invalidates memory cache on its own: https://github.com/SSSD/sssd/blob/master/src/providers/files/files_ops.c#L860 and as you already wrote, files domain have always set expiration time to zero so it does not do anything.

pbrezina commented:
> Lets see what other will say. If your patch will not be merged I will try my best to investigate issue and explain `INTERACTIVE_TIMEOUT/2`

Fair enough. I will submit, two separates PR. One for files provider, one for enumeration. I would advice to push them both so tests are green but if you want to investigate, please do. I certainly will welcome it. But please do it during December so the tests do not keep failing for more years.

pbrezina commented:
* https://github.com/SSSD/sssd/pull/956 Files provider.
* https://github.com/SSSD/sssd/pull/956 Enumeration.

alexey-tikhonov commented:
> > > `poll_canary` has no meaning as we currently block when refreshing the cache. We can either remove it or report a bug against sssd that it behaves incorrectly. @jhrozek When files provider is updating cache in inotify callback (i.e. domain is inconsistent) should nss responder A) return not found and let client look in files database, or B) block until the cache is updated? Currently B is true but tests says A should be done.
> > 
> > Actually... Are you sure "B is true"?
> > I just tested with a patch of `providers/files/files_ops.c:sf_passwd_cb()`:
> > `+ while (1) { usleep(100000); }`
> > right before `sf_enum_files()` and `id` works fine (according to your statement this should block).
> 
> Because you used `usleep` at wrong place. 

I see your point but still I think I used `usleep()` at *right* place. I wished to simulate the case when `sf_enum_files()` blocks (takes very long time) and from my point of view `usleep()` simulates this well.
And thanks to your hint this experiment actually exposed [a bug](https://github.com/SSSD/sssd/pull/955#issuecomment-562114520)

> See, D-Bus communication is asynchronous so once you put a sleep somewhere, the process can not receive nor send any message. In this case, even though `dp_sbus_domain_inconsistent` was called, the message was not send because you immediately go to sleep.
> Therefore, the nss responder will answer with what it has in cache. You will see this in logs.

Indeed. Thanks for pointing this out.

> I'm not sure why it continues with nss_files, I am not this much familiar with nss internals.

If I use `getent passwd` it stops after reading (stale) data from nss_responder.
In case of `id` it continues with `libnss_files`. Strange.
